File: Data Structures & Algorithms/jump-game-ii/submission-1.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        vector<int> dp(nums.size(), 0);
        for(int i = nums.size() - 2; i > -1; --i) {
            int minSteps = 1000;
            if(nums[i] + i == nums.size() - 1) {
                minSteps = 0;
            } else {
                for(int j = i + 1; j <= nums[i] + i && j < nums.size(); ++j) {
                    minSteps = min(minSteps, dp[j]);
                }
            }
            dp[i] = 1 + minSteps;
        }
        // for(auto& i : dp) {
        //     cout << i << " ";
        // }
        return dp[0];
    }
};

File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        candidates = set(nums)
        stack = []
        n = len(nums)
        res = []

        def dfs():
            if len(stack) == n:
                res.append(stack.copy())
                return
            else:
                for candidate in list(candidates):
                    stack.append(candidate)
                    candidates.remove(candidate)
                    dfs()
                    candidates.add(candidate)
                    stack.pop()
        
        dfs()

        return res

                
